logstash-core/lib/logstash/codecs/base.rb

```diff
@@ -28,10 +29,56 @@ def decode(data)
   alias_method :<<, :decode
 
   public
+  # DEPRECATED: Prefer defining encode_sync or multi_encode
   def encode(event)
-    raise "#{self.class}#encode must be overidden"
+    @on_event.call(event, encode_sync(event))
   end # def encode
 
+  # This is a preferred way (along with multiencode) to define a codec's encode function.
+  # This default implementation should really never be invoked by an output plugin
+  # this implementation merely provided for completeness. Plugins should prefer to simply use the
+  # concurrent_multi_receive API which always invokes multi_encode.
+  def encode_sync(event)
+    event, data = multi_encode([event]).first
+    data
+  end
+
+  public
+  # Relies on the codec being synchronous (which they all are!)
+  # We need a better long term design here, but this is an improvement
+  # over the current API for shared plugins
+  # It is best if the codec implements this directly
+  def multi_encode(events)
+    batch = if @has_encode_sync              
+              events.map {|event| [event, self.encode_sync(event)]}
+            else
+              batch = @thread_batches[Thread.current]
+              @multi_encode_mutex.synchronize { batch = @thread_batches[Thread.current] = [] } unless batch
+              batch.clear
+              
+              events.each {|event| self.encode(event) }
+              batch
+            end
+
+    if @on_event != @on_event_proc
+      batch.each {|e| @callback.call(e)}
+    end
+    
+    batch
+  end
+
+  def setup_multi_encode!
+    @has_encode_sync = self.methods.include?(:encode_sync) && self.class.instance_method(:encode_sync) == LogStash::Codecs::Base.instance_method(:encode_sync)
+    @thread_batches = {}
+    @multi_encode_mutex = Mutex.new
+
+    @on_event_proc = proc do |event, data|
+      @thread_batches[Thread.current] << [event, data]
```

Comment:
Thread#[] should be used, imo, instead of a shared-locked hash object who's key is a Thread.

Example:

```
Thread.current[:codec_batches] << [event, data]
```